feat: Remove SSH key write and read user permission PLUTO-630 DOCS-364 (#1765)
codacy/docs: docs/repositories-configure/using-submodules.md

[Git submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) allow you to keep a Git repository as a subdirectory within another Git repository. Git submodules are helpful in maintaining a shared configuration file for your team, and then applying it to multiple Git repositories.

By default, Codacy does normal Git clones that **don't include submodules** to ensure that we only clone necessary repositories. If your organization needs to use submodules, you can request Codacy to enable this feature for you.

## Prerequisites for using submodules

1.  Contact us at <mailto:support@codacy.com> asking to enable submodules on Codacy.

1.  **If you're using Codacy Self-hosted**, [update your license](../chart/maintenance/license.md).

1.  If your submodules are:
    -   **Public repositories**, make sure that your Git URL uses the HTTPS protocol.
    -   **Private repositories**, make sure that your Git URL uses the SSH protocol.

> ## Enabling submodules on a repository {: id="enable"}

Author: @stefanvacareanu7 and @mrfyda
While working on the documentation updates for [removing GitHub SSH write and read user permissions](https://codacy.atlassian.net/browse/PLUTO-624), I needed to touch the procedure to [enable submodules](https://docs.codacy.com/v11.0/repositories-configure/using-submodules/#enabling-submodules), so I took the chance to include [this missing step](https://codacy.atlassian.net/browse/DOCS-364).

Note: As we are removing permissions for GitHub, I assumed the missing step now applies to **Bitbucket and GitHub** (not required for GitLab).

Can you please validate if this procedure is now correct? 🙏 
Thanks!

Reviewer: yes!
the problem is that if you try to add the key you get an error saying that it is already in use

also happens in GitHub
<img width="263" alt="Screenshot 2023-06-29 at 17 50 41" src="https://github.com/codacy/docs/assets/593860/a650c999-2416-4dbc-86b6-7bb0aaa6f004">

> ## Enabling submodules on a repository {: id="enable"}

Reviewer 2: We no longer need the explicit ID for the section. :broom: :nerd_face: 

```suggestion
## Enabling submodules on a repository
```

When using submodules, you must do the following for all your existing and new repositories:

> 1.  Open the repository **Settings**, tab **General**. In the **Danger zone** area, you have the **SSH Key** generated by Codacy to access your repository.

Reviewer 2: The first time I read this it felt weird because it seems like the step is incomplete, since there's no actual thing for the user to do. May be we can ask the user to take note of the SSH key?

Author: Very good point!
It's weird not having an action point in a procedure step. 
I've added your "take note" suggestion as the action point for this step. 👍

    !!! note

>         Codacy generates this repository key when you add a repository to Codacy and uses it to clone that repository. When you're using submodules, Codacy needs to clone additional repositories it may not have access to. To overcome this, Codacy must use an SSH key of your user account to have access to the same repositories as your user.

Reviewer 2: Since this information is now tightly coupled with the step of identifying the SSH key, I suggest that we drop the admonition and simply have the information as a bare paragraph supporting the instruction.

>         Codacy generates this repository key when you add a repository to Codacy and uses it to clone that repository. When you're using submodules, Codacy needs to clone additional repositories it may not have access to. To overcome this, Codacy must use an SSH key of your user account to have access to the same repositories as your user.

Reviewer 2: Since this information is now tightly coupled with the step of identifying the SSH key, I suggest that we drop the admonition and simply have the information as a bare paragraph supporting the instruction.

> 1.  For GitHub and Bitbucket, remove this Codacy key from the repository settings on the Git provider.

Reviewer 2: As a bonus, we could add links to the relevant Git provider docs. But it's probably not worth the effort! :rocket: 

```suggestion
1.  For GitHub and Bitbucket, remove this Codacy key from the repository settings on your Git provider.
```

1.  Add a new SSH key to your git provider account by clicking the link **Add new user key** or the button **Generate New User Key**, depending on your Git provider.

    For GitHub and Bitbucket, this takes you to the Git provider page where you can manage your user account SSH keys. For GitLab, Codacy removes the existing repository key and creates the new SSH key on your user account automatically.

    ![Generate new user key](images/using-submodules-generate-new-user-key.png)

1.  If you're using submodules to share an analysis tool configuration file across your repositories, check if your tool recursively searches the subdirectories of your repositories for configuration files.

    If your tool doesn't detect the configuration files in the submodule directories, you must include a configuration file directly in the root of your repositories referencing the configuration files in the submodule directories.

## Automating user keys for new repositories

!!! info "This section applies only to Codacy Self-hosted"

> You can set Codacy to automatically add the new SSH key to your git provider account for all new repositories by enabling **Add project key to the user, by default** in **Administration**, page **Settings**.

Reviewer 2: Some minor suggestions:

```suggestion
You can set Codacy to automatically add the new SSH key to your Git provider account for all new repositories by enabling **Add project key to the user, by default** on the **Administration** console, page **Settings**.
```

!!! important
    **If you're using Bitbucket Cloud** this setting must be turned off since automatically adding the user keys isn't supported.

![Add project key to the user by default](images/using-submodules-default-add-user-key.png)

## See also

-   [Managing deploy keys in GitHub](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/managing-deploy-keys)
-   [Add an SSH key to your GitHub account](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent)
-   [Configure repository settings in Bitbucket](https://support.atlassian.com/bitbucket-cloud/docs/configure-repository-settings/)

> -   [Add an SSH key to your Bitbucket account](https://support.atlassian.com/bitbucket-cloud/docs/configure-ssh-and-two-step-verification/)

Reviewer 2: It's a great idea to provide these references here, so they don't overcomplicate the instructions above! :ok_hand:

    -   **Public repositories**, make sure that your Git URL uses the HTTPS protocol
    -   **Private repositories**, make sure that your Git URL uses the SSH protocol.

> 1.  If you already have repositories on Codacy where you want to use submodules, add a new [SSH key to your git provider account](#user-account-keys) for each of the repositories. On the repository **Settings** page, you can click the link **Add new user key** to go to your Git provider page where you manage your user account SSH keys.

Reviewer 3: If I understand correctly, the second sentence explains how to achieve what is described in the first one. Some suggestions:

* Minor tweaks to make this relationship more explicit.
* I'd split the last bit and reformat it to appear as a subsection of the instruction step, since it's contextual information and not an explicit instruction.

Thoughts?

```suggestion
1.  If you already have repositories on Codacy where you want to use submodules, [add a new SSH key to your git provider account](#user-account-keys) for each repository by navigating to the repository **Settings** page and clicking the link **Add new user key**.

    This takes you to the Git provider page of the repository, where you can manage your user account SSH keys.

```

Reviewer 2: I really like @nicklem's suggestions to make the instruction more structured. :+1: 

On the other hand, I'd prefer to have the information behind the link to `#user-account-keys` directly here, as a Note admonition and slightly more condensed. For example, I believe that the following sentence is enough to explain why users need to add a Codacy SSH key to their account:

> When you're using submodules, Codacy needs to clone additional repositories it may not have access to. To overcome this, you must add a Codacy SSH key to your user account to grant Codacy access to the same repositories as your user.

Author: Already handled by the [restructured content](https://github.com/codacy/docs/pull/1765#discussion_r1246841017).

    !!! note

>         **If you're using GitLab Cloud**, navigate to the page **Settings** and click the button **Generate New User Key** for each repository. Codacy generates new keys on your user account.

Reviewer 2: This detail seems really "dangling" here just for GitLab. Maybe there's a way to integrate this difference directly in the main instruction?

For example:

> [...] and clicking the link **Add new user key** or the button **Generate New User Key**, depending on your Git provider.
>
> For GitHub and Bitbucket this takes you to the Git provider page of the repository, where you can manage your user account SSH keys. For GitLab, Codacy creates the key on your user account automatically.

## Using submodules on new repositories

> Every time you add new repositories with submodules to Codacy, [repeat steps 2 to 4](#enable) above.

Reviewer 2: This is a convoluted way of providing the instructions. :sweat_smile: (Once again, I'm aware that this were the prior instructions.)

A simpler way of presenting the information would be to have the following sections:

* Pre-requirements (asking Support to enable the feature, update the license for Self-hosted)
* Enabling submodules on a repository (instructions to enable submodules on one repository)
* Automating submodules on new repositories (this section would apply to Self-hosted only, instructions on how to have new repositories automatically use submodules)

Since so few customers use the submodules feature, I don't think it's worthwhile investing in this though. But let me know what you think! :nerd_face:

Author: @prcr I restructured the content following your feedback. 🤓 
I took the chance to add the missing information identified in [DOCS-364](https://codacy.atlassian.net/browse/DOCS-364), as I was already reviewing the repository procedure.

Let me know if it's clear now. 🙏

> Every time you add new repositories with submodules to Codacy, repeat the steps 2 to 4 [above](#enable).

Reviewer 3: I suggest highlighting the sentence fragment with the verb to [frontload the main action](https://codacy.slite.com/app/docs/m4Ynh0i9k8qhqP#ca19f3d3).

```suggestion
Every time you add new repositories with submodules to Codacy, [repeat the steps 2 to 4](#enable) above.
```

**If you're using Codacy Self-hosted**, you can make this the default behavior for all new repositories by enabling **Add project key to the user, by default** in **Administration**, page **Settings**.

!!! important
    **If you're using Bitbucket Cloud** this setting must be turned off since automatically adding the user keys isn't supported.

![Add project key to the user by default](images/using-submodules-default-add-user-key.png)

> ## Why does Codacy need user account SSH keys when using submodules? {: id="user-account-keys"}

Reviewer 3: I like the idea of moving this content to a new section! 👍

When you add a repository to Codacy, Codacy adds a new SSH key to your repository using an integration with your Git provider. Codacy uses this SSH key every time it needs to clone that repository. When you're using submodules, Codacy needs to clone additional repositories it has no access to. Since there is no way of identifying the submodules of a repository before cloning the repository, Codacy adds the SSH keys to the user account so that it has access to the same repositories as the user.

> ## See also

Reviewer 3: This one too! Great UX improvement 👍
